tests/test_services.py

```diff
@@ -0,0 +1,545 @@
+#!/usr/bin/env python3
+# # -*- coding: utf-8 -*-
+
+"""
+Unit tests for Services Commands.
+________________________________________________________________________________
+
+Created by brightSPARK Labs
+www.brightsparklabs.com
+"""
+
+from pathlib import Path
+
+from click.testing import CliRunner
+
+from appcli.commands.configure_cli import ConfigureCli
+from appcli.commands.service_cli import ServiceCli
+from appcli.configuration_manager import ConfigurationManager
+from appcli.logger import enable_debug_logging
+from appcli.models.cli_context import CliContext
+from appcli.models.configuration import Configuration
+from appcli.orchestrators import DockerComposeOrchestrator
+
+# ------------------------------------------------------------------------------
+# CONSTANTS
+# ------------------------------------------------------------------------------
+
+APP_NAME = "test_app"
+
+# directory containing this script
+BASE_DIR = Path(__file__).parent
+
+STACK_CONFIGURATION_FILE = Path(BASE_DIR, "resources/test_stack_settings.yml")
+
+# ------------------------------------------------------------------------------
+# TESTS
+# ------------------------------------------------------------------------------
+
+
+def test_service_start_no_input(tmpdir):
```

Comment:
These tests are running docker containers in the host environment, but are not cleaning up after themselves. This means that tests could unintentionally interfere with each other, or depend on ordering, etc... It also means that the computer you're executing these tests on depends on having 1. docker installed, and 2. the test executor having permissions to run docker commands.


The absolute ideal case for unit tests is to completely isolate all utilised resources between tests (e.g. use of `tmpdir`). This would ideally include the docker runtime environment. I don't think this is going to be particularly easy to do.

Since we don't *require* docker commands to actually be executed (we aren't testing if they succeed/fail, we're testing the orchestrator), I think the best solution here is to mock our process calls. I've only had a really quick look and it should be possible to use a mock/patching framework to mock out the `subprocess.run` and/or `subprocess.Popen` functions as required.

We've used `monkeypatch` previously, but am open to using `unittest.mock` or other frameworks if they're better.

Reply:
By doing this, we should also be able to consolidate all the starts of the tests into a single 'shared' call:
```
    enable_debug_logging()
    config = create_config()
    cli_context = create_cli_context(tmpdir, config)
    config_manager = create_config_manager(cli_context, config)
    config_manager.initialise_configuration()
    config_manager.apply_configuration_changes(message="init apply")
    runner = CliRunner()

    service_command = ServiceCli(config).commands["service"]

```
since we don't actually modify the config